### PR TITLE
Add a util to auto add inverted index

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Add a util to auto add inverted index based on the query response from the table
Add the util into the Pinot tools set
Add a new boolean field 'autoGeneratedInvertedIndex' into IndexingConfig to mark auto-generated invedted index

Mode:
1. NEW: Apply only to tables without inverted index
2. REMOVE: Remove all auto-generated inverted index
3. REFRESH: Refresh the auto-generated inverted index
4. APPEND: Append to the auto-generated inverted index

### CODE REVIEW

**Comment by kishoreg** [comment]
This is awesome Jackie. I suggest moving this class to the controller module. We might want to run this periodically similar to validation manager.

Add a separate command line tool in tools package. Suggest having the following arguments to the tool
- cluster name, zk address, table name or table name pattern //REQUIRED
- min doc threshold required (optional) default can be 100 million.
- avg docs to scan per query assuming all columns are present in the query (allows you to compute the number of columns to add inverted index on) or num of columns to set the inverted index for

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+  private AutoAddInvertedIndex() {
+  }
+
+  private static final String AUTO_GENERATED_INVERTED_INDEX_KEY = "autoGeneratedInvertedIndex";
+
+  private static final String ZK_ADDRESS = "zkAddress:zkPort";
+  private static final String CLUSTER_NAME = "pinot-cluster";
+  private static final String TABLE_CONFIG_PATH = "/CONFIGS/TABLE/";
+  private static final String SCHEMA_PATH = "/SCHEMAS/";
+  private static final ZKHelixAdmin HELIX_ADMIN = new ZKHelixAdmin(ZK_ADDRESS);
```
Comment (kishoreg) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: why do we need to create HELIX_ADMIN here without knowing ZK_ADDRESS

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+    BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(urlConnection.getOutputStream(), "UTF-8"));
+    writer.write(new JSONObject().put("pql", query).toString());
+    writer.flush();
+
+    BufferedReader reader = new BufferedReader(new InputStreamReader(urlConnection.getInputStream(), "UTF-8"));
+    return new JSONObject(reader.readLine());
+  }
+
+  private static class ResultPair implements Comparable<ResultPair> {
```
Comment (kishoreg) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: why not re-use the Immutable Pair from apache commons

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+              // Sort the dimensions based on their cardinalities.
+              Collections.sort(resultPairs);
+
+              // Add the top 2 dimensions into inverted index columns.
+              ResultPair resultPair = resultPairs.get(0);
+              if (resultPair._value > 1) {
+                invertedIndexColumns.add(resultPair._key.substring("distinctCount_".length()));
+              } else {
+                System.err.println("Got abnormal cardinality: " + resultPair + " in table: " + tableName);
```
Comment (kishoreg) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: use logger

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+              continue;
+            }
+
+            // Only add inverted index to table larger than a threshold.
+            JSONObject queryResponse = sendQuery("SELECT COUNT(*) FROM " + tableName);
+            long numTotalDocs = queryResponse.getLong("totalDocs");
+            if (numTotalDocs > TABLE_SIZE_THRESHOLD) {
+
+              // Get each dimension's cardinality on one day's data.
+              queryResponse = sendQuery("SELECT Max(Date) FROM " + tableName);
```
Comment (kishoreg) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: you might want to get the name of time column as input or derive it from the schema.

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  public static void main(String[] args)
+      throws Exception {
+    // Get all resources in cluster.
+    List<String> resourcesInCluster = HELIX_ADMIN.getResourcesInCluster(CLUSTER_NAME);
+    Collections.sort(resourcesInCluster);
+
+    for (String tableName : resourcesInCluster) {
+
+      // Only process on tables with specific keyword.
+      if (tableName.contains("_non_additive_")) {
```
Comment (mcvsubbu) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: take a regex as an argument and match it

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+      return Long.compare(o._value, _value);
+    }
+
+    @Override
+    public String toString() {
+      return _key + ": " + _value;
+    }
+  }
+
+  public static void main(String[] args)
```
Comment (mcvsubbu) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: Invoke it through pinot-admin so that you get args parsing for free? You need to take zk path and cluster name as argument, I suppose.

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+import org.apache.helix.manager.zk.ZNRecordSerializer;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.json.JSONObject;
+
+
+/**
+ * This tool will directly modify the table configs in ZooKeeper.
+ * PLEASE USE IT VERY CAREFULLY!
+ */
+public class AutoAddInvertedIndex {
```
Comment (mcvsubbu) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: The inverted index will not be created immediately, right? We need to "reload" the table. Currently, the servers do not pay attention to a changed table config.

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+  private static final String ZK_ADDRESS = "zkAddress:zkPort";
+  private static final String CLUSTER_NAME = "pinot-cluster";
+  private static final String TABLE_CONFIG_PATH = "/CONFIGS/TABLE/";
+  private static final String SCHEMA_PATH = "/SCHEMAS/";
+  private static final ZKHelixAdmin HELIX_ADMIN = new ZKHelixAdmin(ZK_ADDRESS);
+  private static final ZkHelixPropertyStore<ZNRecord> PROPERTY_STORE =
+      new ZkHelixPropertyStore<>(ZK_ADDRESS, new ZNRecordSerializer(),
+          PropertyPathConfig.getPath(PropertyType.PROPERTYSTORE, CLUSTER_NAME));
+
+  private static final String BROKER_QUERY_URL = "http://brokerAddress:brokerPort/query";
```
Comment (antumbde) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: incorrect for final string

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+import org.apache.helix.PropertyType;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.manager.zk.ZKHelixAdmin;
+import org.apache.helix.manager.zk.ZNRecordSerializer;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+import org.json.JSONObject;
+
+
+/**
+ * This tool will directly modify the table configs in ZooKeeper.
```
Comment (antumbde) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: It shouldn't. Use controller APIs to update table configuration

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+ */
+public class AutoAddInvertedIndex {
+  private AutoAddInvertedIndex() {
+  }
+
+  private static final String AUTO_GENERATED_INVERTED_INDEX_KEY = "autoGeneratedInvertedIndex";
+
+  private static final String ZK_ADDRESS = "zkAddress:zkPort";
+  private static final String CLUSTER_NAME = "pinot-cluster";
+  private static final String TABLE_CONFIG_PATH = "/CONFIGS/TABLE/";
```
Comment (antumbde) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: This command needs whole bunch of parameters for usability

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+  private static final String SCHEMA_PATH = "/SCHEMAS/";
+  private static final ZKHelixAdmin HELIX_ADMIN = new ZKHelixAdmin(ZK_ADDRESS);
+  private static final ZkHelixPropertyStore<ZNRecord> PROPERTY_STORE =
+      new ZkHelixPropertyStore<>(ZK_ADDRESS, new ZNRecordSerializer(),
+          PropertyPathConfig.getPath(PropertyType.PROPERTYSTORE, CLUSTER_NAME));
+
+  private static final String BROKER_QUERY_URL = "http://brokerAddress:brokerPort/query";
+
+  private static final long TABLE_SIZE_THRESHOLD = 100_000_000;
```
Comment (antumbde) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: provide unit on the size

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+
+  private static AbstractTableConfig getTableConfig(String tableName)
+      throws Exception {
+    return AbstractTableConfig.fromZnRecord(PROPERTY_STORE.get(TABLE_CONFIG_PATH + tableName, null, 0));
+  }
+
+  private static void setTableConfig(String tableName, AbstractTableConfig tableConfig)
```
Comment (antumbde) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: use controller APIs

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java (hunk 4 above)
Comment (antumbde) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: FYI, servers also have API to provide full metadata through REST apis

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java (hunk 5 above)
Comment (antumbde) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: Take tableName as param as Subbu suggested. 
Also if (! match) { continue} to reduce deep nesting levels

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+        }
+
+        // Remove empty string.
+        int emptyStringIndex;
+        while ((emptyStringIndex = invertedIndexColumns.indexOf("")) != -1) {
+          invertedIndexColumns.remove(emptyStringIndex);
+        }
+
+        // Only add inverted index if there is no existing ones.
+        if (invertedIndexColumns.isEmpty()) {
```
Comment (antumbde) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: suggestion: invert condition to reduce nesting level

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+          // Table must have a schema to add inverted index.
```
Comment (antumbde) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: why is this required ? I thought offline tables don't need schema

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+            if (numTotalDocs > TABLE_SIZE_THRESHOLD) {
+
+              // Get each dimension's cardinality on one day's data.
+              queryResponse = sendQuery("SELECT Max(Date) FROM " + tableName);
+              int maxDate = queryResponse.getJSONArray("aggregationResults").getJSONObject(0).getInt("value");
+              // Query DISTINCTCOUNT on all dimensions in one query might cause timeout, so query them separately.
+              List<ResultPair> resultPairs = new ArrayList<>();
+              for (String dimensionName : dimensionNames) {
+                String query =
+                    "SELECT DISTINCTCOUNT(" + dimensionName + ") FROM " + tableName + " WHERE Date = " + maxDate;
```
Comment (antumbde) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: distinctcount on dimension name can be significant hit (date predict may soften it but not sure...)

Use segment metadata api to get cardinality directly

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java (hunk 7 above)
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: Right. This tool will only modify the table config.

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java (hunk 1 above)
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: ZK_ADDRESS is hardcoded for this initial version. I will make it an argument passed in by the user.

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java (hunk 2 above)
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: ImmutablePair is a final class. I want to make this pair comparable, and override the toString() method to print useful information.

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java (hunk 4 above)
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: Will make it general.

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java (hunk 6 above)
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: This is an initial version. Will make it general and a command in pinot-admin.

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java (hunk 9 above)
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: I will also add "autoGeneratedInvertedIndex" flag into ZNRecord, so for now I need to modify in ZooKeeper.

File: pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java (hunk 10 above)
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/com/linkedin/pinot/tools/AutoAddInvertedIndex.java]: I agree. Will work on turning this into a usable command in pinot-admin.

**Comment by Jackie-Jiang** [comment]
@kishoreg @mcvsubbu @antumbde Refactored the code to make it more general and suitable to use as a tool or controller service.
Configurable arguments: tableNamePattern, tableSizeThreshold, cardinalityThreshold, maxNumInvertedIndex
The current algorithm still use DISTINCTCOUNT to get the cardinality. We can plug in other strategies later.

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+  private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
+  private final Strategy _strategy;
+  private final Mode _mode;
+
+  private String _tableNamePattern = null;
+  private long _tableSizeThreshold = DEFAULT_TABLE_SIZE_THRESHOLD;
+  private long _cardinalityThreshold = DEFAULT_CARDINALITY_THRESHOLD;
+  private int _maxNumInvertedIndexAdded = DEFAULT_MAX_NUM_INVERTED_INDEX_ADDED;
+
+  public AutoAddInvertedIndex(@Nonnull String zkAddress, @Nonnull String clusterName, @Nonnull String controllerAddress,
```
Comment (jfim) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java]: Use `PinotHelixResourceManager` instead of interacting with `ZKHelixAdmin` directly.

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+
+      // Get the inverted index config
+      AbstractTableConfig tableConfig = getTableConfig(tableName);
+      IndexingConfig indexingConfig = tableConfig.getIndexingConfig();
+      List<String> invertedIndexColumns = indexingConfig.getInvertedIndexColumns();
+      boolean autoGeneratedInvertedIndex = indexingConfig.isAutoGeneratedInvertedIndex();
+
+      // Handle auto-generated inverted index
+      if (autoGeneratedInvertedIndex) {
+        assert !invertedIndexColumns.isEmpty();
```
Comment (jfim) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java]: Use Guava preconditions instead of Java asserts (Java asserts can be turned off)

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+        continue;
+      }
+      String timeColumnName = timeFieldSpec.getName();
+
+      // Only add inverted index to table larger than a threshold
+      JSONObject queryResponse = sendQuery("SELECT COUNT(*) FROM " + tableName);
+      long numTotalDocs = queryResponse.getLong("totalDocs");
+      LOGGER.info("Table: {}, number of total documents: {}", tableName, numTotalDocs);
+      if (numTotalDocs <= _tableSizeThreshold) {
+        LOGGER.info("Table: {}, skip adding inverted index because too less total documents: {}", tableName,
```
Comment (jfim) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java]: "because the table is too small"

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java [hunk 19]
```diff
@@ -0,0 +1,360 @@
...
+            numTotalDocs);
+        continue;
+      }
+
+      // Get each dimension's cardinality on one day's data
+      queryResponse = sendQuery("SELECT Max(" + timeColumnName + ") FROM " + tableName);
+      int maxTimeStamp = queryResponse.getJSONArray("aggregationResults").getJSONObject(0).getInt("value");
+      LOGGER.info("Table: {}, max time column {}: {}", tableName, timeColumnName, maxTimeStamp);
```
Comment (jfim) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java]: Maybe emit a warning/error out if the time granularity is not days

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java [hunk 20]
```diff
@@ -0,0 +1,360 @@
...
+    writer.flush();
+
+    BufferedReader reader = new BufferedReader(new InputStreamReader(urlConnection.getInputStream(), "UTF-8"));
+    return new JSONObject(reader.readLine());
+  }
+
+  private boolean updateIndexConfig(String tableName, AbstractTableConfig tableConfig)
+      throws Exception {
+    HttpURLConnection httpURLConnection = (HttpURLConnection) new URL(
+        "http://" + _controllerAddress + "/tables/" + tableName + "/indexingConfigs").openConnection();
```
Comment (jfim) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java]: Use `ControllerRequestURLBuilder` to build REST API URLs. In this case, you'll need to add the one to upload the indexing configs.

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java (hunk 16 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java]: Because PinotHelixResourceManager might create a new controller, for now I will use ZKHelixAdmin. When we actually add this util into controller in the future, we can add a new constructor with PinotHelixResourceManager.

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java (hunk 17 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java]: fixed.

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java (hunk 18 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java]: fixed.

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java (hunk 19 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java]: Added a warning.

File: pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java (hunk 20 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/util/AutoAddInvertedIndex.java]: Fixed.
